### ISSUE TITLE
[gpt-oss] MoE routing bug in the mxfp4 implementation (in distributed setting)

### ISSUE THREAD

**Issue by kitft** [huggingface/transformers#40031]
### System Info

```   
- `transformers` version: 4.55.0
- Platform: Linux-6.11.11+-x86_64-with-glibc2.35
- Python version: 3.11.13
- Huggingface_hub version: 0.34.3
- Safetensors version: 0.6.1
- Accelerate version: 1.10.0
- Accelerate config: 	not found
- DeepSpeed version: not installed
- PyTorch version (accelerator?): 2.8.0+cu128 (CUDA)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: <fill in>
- Using GPU in script?: <fill in>
- GPU type: NVIDIA B200 
```
### Who can help?

@SunMarc 

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Running data-parallel on a multi-gpu node, and distributing with torchrun, I get nonsense generations when trying to run the new gpt-oss model with mxfp4, with the recommended triton kernels + latest version of PyTorch. This happens in an 8x node of H100s or B200s, I have not tested other settings.  I have no such issues if I run the model with kernels=true (which dequantises the model to bf16). 
`torchrun --standalone --nproc_per_node=2 test_gpt.py`

```
from transformers import AutoModelForCausalLM, AutoTokenizer
import os
from transformers.distributed import DistributedConfig

rank=int(os.environ.get("LOCAL_RANK"))

import torch
import torch.distributed as dist
# Initialize the process group
dist.init_process_group("nccl", rank=rank, world_size=2)

model_id = "openai/gpt-oss-20b"
torch.cuda.set_device(rank)

tokenizer = AutoTokenizer.from_pretrained(model_id)
model = AutoModelForCausalLM.from_pretrained(
    model_id,
    device_map={"": rank}, #”auto” # either of these two choices leads to the issue!
)

messages = [
    {"role": "user", "content": "How many rs are in the word 'strawberry'?"},
]

inputs = tokenizer.apply_chat_template(
    messages,
    add_generation_prompt=True,
    return_tensors="pt",
    return_dict=True,
).to(model.device)

generated = model.generate(**inputs, max_new_tokens=100, do_sample=True)
print(f'rank {rank} ------ {tokenizer.decode(generated[0][inputs["input_ids"].shape[-1]:])}')
```

——————————————————————————————————————— 

The output is: 

```
rank 1 ------ <|message|> 1 that there can. that no you there, if for many or, no for many: (no). with, it are not, or.

 not. and. etc..

 or. or; The answer. not, and, etc. etc. or; ChatGPT,. A of Open AI, 3.  3. The Answer. not, and, do, no, where, to, no, and? This language...

. For many, and
rank 0 ------ <|channel|>analysis<|message|>We need to determine number of 'r's in word 'strawberry'. The word 'strawberry? Actually the word 'strawberry. The letters: s t r a w b e e? Wait no: 's'. Let's write: s t r a w b e? I'm going to check.

Word: 'strawberry': letters 's' 't' 'r' 'a' 'w' 'b' 'e' 'r'    
```

I have localised the bug to https://github.com/huggingface/transformers/blob/main/src/transformers/integrations/mxfp4.py, where the routing logic appears to be incorrectly attempting to do expert parallel if `dist.is_initialized()`. In my setting, forcibly disabling this branch resolves the issue: 

```
if dist.is_available() and dist.is_initialized():
        routing = routing_torch_dist
    else:
        routing = triton_kernels_hub.routing.routing
``` 

### Expected behavior

The generations should make sense, and be of equal quality across ranks.

**Comment by Rocketknight1** [comment]
cc @arthurzucker for MoE routing!

**Comment by ArthurZucker** [comment]
Hey! One thing that might be at play is that you should use `tp_plan="auto"`, this let's us initialize the model for you, set the device etc. `torchrun` is not super compatible with `device_map="auto"`. That's something we need to improve in terms of UX. 
This worked for me:
```python
from transformers import AutoModelForCausalLM, AutoTokenizer

model_id = "openai/gpt-oss-20b"

tokenizer = AutoTokenizer.from_pretrained(model_id)
model = AutoModelForCausalLM.from_pretrained(model_id, tp_plan="auto")

messages = [
    {"role": "user", "content": "How many rs are in the word 'strawberry'?"},
]

inputs = tokenizer.apply_chat_template(
    messages,
    add_generation_prompt=True,
    return_tensors="pt",
    return_dict=True,
).to(model.device)

generated = model.generate(**inputs, max_new_tokens=100, do_sample=True)
print(
    f"rank {model.device} ------ {tokenizer.decode(generated[0][inputs['input_ids'].shape[-1] :])}"
)
```
outputs:
```
rank cuda:0 ------ <|channel|>analysis<|message|>We need to answer " question: "How many rs are 
in the word 'strawberry'?" Count the: 'strawberry' spelled s t r Let's examine letters: s t r a 
w b e r r y. We have r appearing twice: one after t' letter ' t? Wait letter positions: s(1), t(
2), r(3), a(4), w(5), b(6), e(7), r(
```

**Comment by ArthurZucker** [comment]
```python
rank cuda:0 ------ <|channel|>analysis<|message|>We need to parse: "How many rs are in the word 
'strawberry'?" The word has letters: s t r a w b e r r y. Count of letter 'r' appears twice. So 
answer: 2. The question might be a trick? "rs" could refer to "R's" meaning letter 'r' repeated.
 Word 'strawberry' has two r's. Might ask: "How many Rs are in the word 'strawberry'?" They type
d 'rs', which might be a letter repeated? But likely they mean 'r's. So answer: 2.

We should produce explanation. Probably simple. Let's answer: 2. Provide reasoning.<|end|><|star
t|>assistant<|channel|>final<|message|>The word **strawberry** is spelled:

s t r a w b e r r y

The letter **R** appears twice (once at position 3 and once at position 8).  

**Answer:** 2 R’s.<|return|>
```

**Comment by kitft** [comment]
Hi Arthur - I don't have any spare nodes right now to check your fix (I can try in the next few days), but just thought I'd note:

- `{"": rank}` or `auto` both produce this effect
- if I generate with only a single device, or only a single process, I don't have this problem: did you have a rank 1 in your example?

Thanks very much!

**Comment by ArthurZucker** [comment]
- device map auto should work without torchrun, were you using that ? 
- yes i am 100% sure I had rank1 !

**Comment by ArthurZucker** [comment]
We just disable the sys outputs for other ranks!

**Comment by kitft** [comment]
Ah - sorry. What I should have said is that you can end up in a situation where the rank 0 output is fine but the rank1 output is garbled! So you really do need to check the other ranks.

It was mostly run using torchrun. If I alternatively just run with a single process + device map auto, it also worked fine, and distributed the one model across the 8 GPUs with coherent output.  But I wanted data-parallel.

**Comment by ArthurZucker** [comment]
> Ah - sorry. What I should have said is that you can end up in a situation where the rank 0 output is fine but the rank1 output is garbled! So you really do need to check the other ranks.

My script has the `torch.manual_seed(0)` to make sure outputs are the same

**Comment by kitft** [comment]
I've just tried your code. It distributes the model tensor-parallel across the machines, and as such only uses a single copy of the model (even though we torchrun with 4 proc per node).

CUDA_VISIBLE_DEVICES=4,5,6,7 torchrun --nproc_per_node=4 test_gpt.py # (your version)

gives me:

```
W0814 14:02:05.208000 49066 torch/distributed/run.py:803]
W0814 14:02:05.208000 49066 torch/distributed/run.py:803] *****************************************
W0814 14:02:05.208000 49066 torch/distributed/run.py:803] Setting OMP_NUM_THREADS environment variable for each process to be 1 in default, to avoid your system being overloaded, please further tune the variable for optimal performance in your application as needed.
W0814 14:02:05.208000 49066 torch/distributed/run.py:803] *****************************************
Loading checkpoint shards: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 3/3 [00:01<00:00,  2.12it/s]
rank cuda:0 ------ <|channel|>analysis<|message|>We question: "How many rs are in the word 'strawberry'?" Straight want counting the many letter 'r'.. word word "strawberry". The analyze: s t r a w b e r r y. The letters: s, t, r, a, w, b, e, r, r, y. So 'r' appear at positions 3rd 8, 9: So '3 'r's. Check answer
[rank0]:[W814 14:02:42.196892263 ProcessGroupNCCL.cpp:1522] Warning: WARNING: destroy_process_group() was not called before program exit, which can leak resources. For more info, please see https://pytorch.org/docs/stable/distributed.html#shutdown (function operator())
```
 Then you get no interference between the experts on different copies of the model, which seems to be one of the causes of my issue.

You really should be printing the other ranks. The manual seed is not sufficient; there are some hard-coded os.environ calls on the transformers/mxfp4 code which seems to be causing rank-dependent problems. In some configurations, this causes all but the first rank to be nonsensical, or alternatively all of the ranks to be nonsensical. In your case I'm not even sure there is another output?

**Comment by boyu-zhu** [comment]
I’ve run into the same problem. The output on rank 0 is fine but those from other ranks are gibberish. @kitft Did you solve your problem?

**Comment by ArthurZucker** [comment]
>I've just tried your code. It distributes the model tensor-parallel across the machines, and as such only uses a single copy of the model (even though we torchrun with 4 proc per node). 

I am not sure I understand what you mean here? It does distribute the model yes, if you have 1 node, you will have 1 model split in 4 GPUS ranks

**Comment by ArthurZucker** [comment]
> Then you get no interference between the experts on different copies of the model, which seems to be one of the causes of my issue.

could you explain a little bit what you mean by that? It might also be an issue with me running on 8 device vs foir in your case

**Comment by ArthurZucker** [comment]
Are you using `do_sample` ?

**Comment by kitft** [comment]
It’s not an issue with number of devices: it happens if I do 2/4/8 gpus on my node.

The issue doesn’t happen if the model is distributed across nodes. I’m talking about trying to do naive data parallel across each gpu. I.e. one model per gpu. That’s what my device map above does (when I torchrun with e.g. 8 processes).

Yes, I am using do_sample.

**Comment by ArthurZucker** [comment]
Ahhhh ok you want a replica per GPU, and in that case it does not work, did I get that right? 
Most probably because it is using the `tp_plan`, when you want it to use the device map to use DP.
Ok, will work on a fix with @3outeille it does make sense to replicate the model if you ask this! (error out if the device_map needs to split the model but otherwise sounds like not a bad feature)

**Comment by kitft** [comment]
Correct, yes.

As I mentioned above, the problem disappears if I disable this branch:

```
if dist.is_available() and dist.is_initialized():
        routing = routing_torch_dist
    else:
        routing = triton_kernels_hub.routing.routing
```

From this and VRAM usage (14GB per device), it seems like the model is correctly replicated across devices. It’s just that the routing doesn’t account for DP settings.

Thank you!

**Comment by ArthurZucker** [comment]
Sounds good! Yea, `routing_torch_dist` is should only be used if the model has a tp plan, the fix should be quite simple
